DO NOT MERGE gpus <- mpi

bmcdanie opened the pull request.

bmcdanie committed:
First pass

bmcdanie committed:
Optional openmp for parallel batch build

bmcdanie committed:
Interface for distribution plan component

bmcdanie committed:
Add mapping from assigned chunks to input/output vectors

bmcdanie committed:
task splitting function

bmcdanie committed:
working

bmcdanie committed:
draft element grid division

bmcdanie committed:
working

bmcdanie committed:
working

bmcdanie committed:
working - move back to fusiont5

bmcdanie committed:
add combine function for subgrid

bmcdanie committed:
working - squash to here

bmcdanie committed:
Test subgrid struct

bmcdanie committed:
Add distribution tests and correct identified errors

bmcdanie committed:
Finish initial distribution tests

bmcdanie committed:
Fix compile warnings

bmcdanie committed:
Add reduction partner finding code

bmcdanie committed:
Finish partial combine dimensions w/ tests

bmcdanie committed:
Extend chunk tests for subgrids.

Fix the Q#@#*(^$&*(^ local row bug

bmcdanie committed:
Add optional mpi build

bmcdanie committed:
Fix MPI build, add local rank function

bmcdanie committed:
cleanup for review

bmcdanie committed:
wrap couts in main to print once per node

dlg0 commented on src/chunk.cpp:
```diff
@@ -22,7 +22,7 @@ int max_connected_in_chunk(element_chunk const &g)
 }
 
-limits columns_in_chunk(element_chunk const &g)
+limits<> columns_in_chunk(element_chunk const &g)
```
Is this just using the default template arguments? (did it even compile before without the empty chevrons?)

dlg0 commented on src/chunk.cpp:
```diff
@@ -176,6 +176,35 @@ int get_num_chunks(element_table const &table, PDE<P> const &pde,
 }
 
+// determine how many chunks will be required to solve the problem
+// a chunk is a subset of the element subgrid whose total workspace requirement
```
We are going to need a drawing or something to explain how chunks relate to the element subgrid relate to the worklist relate to the actual matrix :)

bmcdanie committed:
make combine dimensions (partial) more flexible

bmcdanie committed:
Fix statically allocated device tensors issue.

Include coefficient matrices in chunk size calculation

bmcdanie committed:
use new subgrid functions in main

bmcdanie committed:
set correct device for local rank

bmcdanie committed:
Running on two GPUS

bmcdanie committed:
working

bmcdanie committed:
working

bmcdanie committed:
working

bmcdanie committed:
working

bmcdanie committed:
Fix x-indexing in batch build for subgrids

bmcdanie committed:
commit to share w/ harry

stonecoldhughes committed:
Added new component to PDE spec heirarchy.

"class term" objects now contain "partial_term" objects
that are chained together

bmcdanie committed:
Finish reduction communication tests

bmcdanie committed:
Add logic to deal with unused processes

bmcdanie committed:
Fix for non-gpu case

stonecoldhughes committed:
partial_terms store partial coefficient matrices from g_funcs

bmcdanie committed:
Trim unused partnering function

bmcdanie committed:
Partnering technique

bmcdanie committed:
working

bmcdanie committed:
refactor row space intervals code

bmcdanie committed:
working

bmcdanie committed:
Add mpi send/recv

bmcdanie committed:
fix mpi send/recv to use correct buffer

bmcdanie committed:
fix rs intervals

bmcdanie committed:
Gather errors from all ranks for printing

bmcdanie committed:
Reassemble solution vector

bmcdanie committed:
Use send/recv to avoid overflow displacements

bmcdanie committed:
Cleanup

bmcdanie committed:
working

bmcdanie committed:
update chunk tests for grids/refactor

bmcdanie committed:
Refactor and update time advance tests

bmcdanie committed:
working

bmcdanie committed:
working

bmcdanie committed:
Fix time advance test problem caused by wrong cfl

bmcdanie committed:
finish time advance tests

bmcdanie committed:
working

bmcdanie committed:
Finish test update/refactor

bmcdanie committed:
Correct errors when building with MPI=0

bmcdanie committed:
Add tests for two more distribution functions

bmcdanie committed:
finish distribution tests

bmcdanie committed:
Finish rebase

bmcdanie committed:
Pull in harry's changes and refactor

bmcdanie committed:
rename for clarity

stonecoldhughes committed:
transform from wavelet space to real space with tests and gold data.

dlg0 committed:
wip on fokkerplanck2_complete

bmcdanie committed:
integrate mpi and distribution components

bmcdanie committed:
complete distribution tests0

bmcdanie committed:
fix compiler warning

bmcdanie committed:
cleanup

dlg0 committed:
wip - but running with 1 term

dlg0 committed:
wip - all 7 terms added and running

dlg0 committed:
match get_dt to matlab

dlg0 committed:
fokkerplanck2_complete testing passing for pde and coefficients - still to do time_advance

dlg0 committed:
fokkerplanck2_complete done (explicit anyway)

bmcdanie committed:
use c-style mpi types

bmcdanie committed:
Fix openmp/mkl threading framework conflict

bmcdanie committed:
Fix test for clang - can't deadlock on messages to myself

stonecoldhughes committed:
Described PDEs with partial terms.

grahamlopez committed:
Merge pull request #136 from bmcdanie/feature/parallel_batch

Feature/parallel batch (depends on feature/gpus)

stonecoldhughes committed:
Moved coefficient generation into separate function.
Changed g_func_type to all double precision.
Updated all relevant tests to use new coefficient generation function.
Refactored and renamed variables.

bmcdanie committed:
use standard build macro technique

bmcdanie committed:
fix review items

bmcdanie committed:
Trim unused function

bmcdanie committed:
clang-format...grr

bmcdanie committed:
address graham review comment

bmcdanie committed:
Add asserts

bmcdanie committed:
remove clang-format tags

grahamlopez committed:
Merge pull request #153 from stonecoldhughes/refactor/pde_spec

Refactor/pde spec

bmcdanie committed:
format

bmcdanie committed:
Fix TA test error in refactor.

Ensure we aren't trying to use unimplmented gpu implicit

bmcdanie committed:
Rebased fp2

bmcdanie committed:
remove unused switch cases

bmcdanie committed:
Fix broken assert

bmcdanie committed:
fix const correctness

bmcdanie committed:
remove clang-format off/on

bmcdanie committed:
move function in tests general, east const

bmcdanie committed:
Remove vesitigal files

bmcdanie committed:
first round of MR review fixes

bmcdanie committed:
these should be pass-by-ref...

bmcdanie committed:
Commit for benchmark runs

bmcdanie committed:
Catch another pass-by-value

bmcdanie committed:
more MR comments

bmcdanie committed:
clangformat...

bmcdanie committed:
Explicitly declare move and copy for message

bmcdanie committed:
store coefficient device matrices in pde

bmcdanie committed:
Another round of MR review items

bmcdanie committed:
address review comments

bmcdanie committed:
rename function, default copy/move for messages

bmcdanie committed:
Implement MPI testing for distribution / time advance components

bmcdanie committed:
Fix non-mpi build issue

bmcdanie committed:
Fix cuda build error

bmcdanie committed:
Fix options test to work when distributed

bmcdanie committed:
add const

bmcdanie committed:
Workaround for missing include in highfive

grahamlopez committed:
Merge pull request #180 from bmcdanie/fix/h5include

Workaround for missing include in highfive

Commit:
squash - does not compile. Integrating realspace transform into main.

bmcdanie committed:
Realspace output

bmcdanie committed:
complete integration into main

bmcdanie committed:
whoops

bmcdanie committed:
Rename

bmcdanie committed:
follow standard generated inputs structure

grahamlopez committed:
Merge pull request #164 from dlg0/fokkerplank2_complete

fokkerplank2_complete (depends on 153)

grahamlopez committed:
Merge pull request #173 from stonecoldhughes/feature/wavelet_to_real

transform from wavelet space to real space with tests

bmcdanie committed:
fix compiler warnings...

grahamlopez committed:
Merge pull request #181 from bmcdanie/fix/compiler-warn

fix compiler warnings...

bmcdanie committed:
Disable realspace transform when MPI used. cleanup tests

bmcdanie committed:
Exit on bad parse

bmcdanie committed:
reorder

Commit:
allow overriding testing rank numbers
